s32/CMakeLists.txt

```diff
@@ -22,3 +22,32 @@ zephyr_compile_definitions(${S32_DEVICE})
 
 add_subdirectory(drivers/${DRIVERS_BASE})
 add_subdirectory(soc/${SOC_BASE})
+
+# Some shim drivers use Kconfig string options to glue with driver configuration.
```

Comment:
How/where is s32ct_define.h used?

Reply:
Glancing at how this is getting used, we still seem to have "BOARD" bits in `s32/soc/s32z27/src/Linflexd_Uart_Ip_Sa_BOARD_InitPeripherals_PBcfg.c`.   Can we not just have instances of  `Linflexd_Uart_Ip_UserConfigType` setup in the shim driver.  What happens when someone wants needs a second or third UART?

Reply:
> How/where is s32ct_define.h used?

I have pushed the following changes:
- renamed the generated header from `s32ct_define.h` to `s32ct_conf.h`, which I consider more appropriate
- removed the global include through imacros of (now) `s32ct_conf.h` and include this header in each shim driver source as needed (e.g. https://github.com/zephyrproject-rtos/zephyr/pull/49161/commits/245876f67ed631f772c4040365aef54524815e9b#diff-210804e4917b7166b3e91bff5b45853537b6468078848fa45bdcc8b6bf5d4499R10)

At the moment this auto-generated header is used only by the serial driver, to build a macro with the name of the default configuration for Uart (https://github.com/zephyrproject-rtos/zephyr/pull/49161/commits/245876f67ed631f772c4040365aef54524815e9b#diff-210804e4917b7166b3e91bff5b45853537b6468078848fa45bdcc8b6bf5d4499R332). The use is not limited to this case, but other peripheral drivers (that we plan on open a PR after this one) make use of it as well.

Reply:
> Glancing at how this is getting used, we still seem to have "BOARD" bits in s32/soc/s32z27/src/Linflexd_Uart_Ip_Sa_BOARD_InitPeripherals_PBcfg.c

The "BOARD_InitPeripherals" is just a default name set by the configuration tool. The code generated doesn't depend on the Zephyr board being built. So future boards based on the same SoC can still use the generated code.

> Can we not just have instances of Linflexd_Uart_Ip_UserConfigType setup in the shim driver. What happens when someone wants needs a second or third UART?

I have reworked the integration between the SDK drivers, their default configuration and Zephyr shim drivers, to align better with the approach followed by other HAL's and remove the need for the user to use the configurator tools to enable other instances than the ones available in the default configuration.

The SDK driver default configuration serves now as template to populate at run-time (during device initialization) the init configuration for the peripheral instances enabled through devicetree. For example, https://github.com/zephyrproject-rtos/zephyr/pull/49161/commits/178eaf526754841d72a5d7e8dfbbaa0d1d22d4fd#diff-210804e4917b7166b3e91bff5b45853537b6468078848fa45bdcc8b6bf5d4499R296-R300. In the case of UART, the instance configuration must be stored in the device data struct because the SDK driver will keep a reference to it. This may not be required for other peripheral drivers, and the instance configuration could live in the stack of the device init function to save memory.

> How/where is s32ct_define.h used?

Due to the rework, this auto-generated header and the script generating it, are no longer necessary.

Reply:
I still dont see how or why we need `Linflexd_Uart_Ip_xHwConfigPB_0_BOARD_INITPERIPHERALS` and thus `Linflexd_Uart_Ip_Sa_BOARD_InitPeripherals_PBcfg.c`.  It seems like how `Linflexd_Uart_Ip_xHwConfigPB_0_BOARD_INITPERIPHERALS` is defined will ONLY work for UART0, since it has `&Linflexd_Uart_Ip_apStateStructure[0]`.  Doesn't that need to change based on the UART being used?

Reply:
`Linflexd_Uart_Ip_Sa_BOARD_InitPeripherals_PBcfg.c` is a file generated by the configuration tool and contains initialization code and state structures for the UART peripheral. `Linflexd_Uart_Ip_xHwConfigPB_0_BOARD_INITPERIPHERALS` is the struct used to initialize UART instance 0, but we are using it as a "template" for the initial configuration of all UART instances. The behavior is _somehow analogous_ to the `*_GetDefaultConfig()` functions used by MCUX to populate the peripheral initialization struct. The UART shim driver uses this template struct with default configuration to create its own private copy (per device instance), modifying the struct fields as needed. The only modification to the default configuration being done at the moment, is to update the reference to the UART state struct (`&Linflexd_Uart_Ip_apStateStructure[0]`) based on the device instance initialized. We plan in the next PR's to introduce the ability to modify more fields through devicetree properties. You can see current implementation in https://github.com/zephyrproject-rtos/zephyr/blob/178eaf526754841d72a5d7e8dfbbaa0d1d22d4fd/drivers/serial/uart_s32_linflexd.c#L296-L300.

The tool-generated code for UART is rather simple, but for other peripherals that we plan to support, the generated code is more complex (e.g. peripherals that interact with each other) and there are a few configuration structs per instance. This is one of the reasons we have kept `Linflexd_Uart_Ip_Sa_BOARD_InitPeripherals_PBcfg.c` as it is generated by the tool for the time being. Another reason is that is going to be easier do updates of the tool-generated code. Nevertheless, after the feedback on this PR we will continue exploring a better approach to perhaps depend less on the configuration tool and overcome any shortcoming of current implementation.

Reply:
W/o seeing what generated code for the other peripherals looks like its hard to comment on them.  I'd say this is different from `*_GetDefaultConfig()` since that isn't generated and static.  I'd suggest it makes more sense to just setup `Linflexd_Uart_Ip_xHwConfigPB_0_BOARD_INITPERIPHERALS` in the shim driver and not use the generated code as a template for the UART.  We can cross the bridge as we come to it for other peripherals as those are supported.

Reply:
I think thats it at this point.